packages/osd-pm/src/utils/validate_dependencies.ts

```diff
@@ -138,21 +182,181 @@ export async function validateDependencies(osd: OpenSearchDashboards, yarnLock:
     }
   }
 
-  const duplicateRanges = Array.from(depRanges.entries())
-    .filter(([, ranges]) => ranges.length > 1)
-    .reduce(
-      (acc: string[], [dep, ranges]) => [
-        ...acc,
-        dep,
-        ...ranges.map(
-          ({ range, projects }) => `  ${range} => ${projects.map((p) => p.name).join(', ')}`
-        ),
-      ],
-      []
-    )
-    .join('\n        ');
-
-  if (duplicateRanges) {
+  const cachedManifests = new Map<string, any>();
+  const violatingSingleVersionDepRanges = new Map<
+    string,
+    Array<{ range: string; projects: Project[] }>
+  >();
+  depRangesLoop: for (const [depName, ranges] of depRanges) {
+    // No violation if just a single range of a dependency is used
+    if (ranges.length === 1) continue;
+
+    const installedVersions = new Set<string>();
+    const installedDepVersionsCache = new Map<string, string>();
+    const desiredRanges = new Map<string, Project[]>();
+
+    rangesLoop: for (const { range, projects } of ranges) {
+      for (const project of projects) {
+        if (!cachedManifests.has(project.path))
+          cachedManifests.set(
+            project.path,
+            // If there are errors reading or parsing the lockfiles, don't catch and let them fall through
+            parseLockFile(readFileSync(path.join(project.path, 'yarn.lock'), 'utf-8'))
+          );
+        const { object: deps } = cachedManifests.get(project.path);
+        if (deps?.[`${depName}@${range}`]?.version) {
+          installedVersions.add(deps[`${depName}@${range}`].version);
+          installedDepVersionsCache.set(
+            `${project.name}#${depName}`,
+            deps[`${depName}@${range}`].version
+          );
+        } else {
+          log.warning(`Failed to find the installed version for ${depName}@${range}`);
+          // If we cannot read any one of the installed versions of a depName, there is no point in continuing with it
+          installedVersions.clear();
+          desiredRanges.clear();
+          break rangesLoop;
+        }
+      }
+
+      desiredRanges.set(range, projects);
+    }
+
+    // More than one range is used but couldn't get all the installed versions: call out violation
+    if (installedVersions.size === 0) {
```

Comment:
If `installedVersions.size` equal 1, shall we check if the installed versions was from different ranges?